src/kubernetes_charm.py

```diff
@@ -145,6 +156,42 @@ def _patch_service(self) -> None:
     #  Handlers
     # =======================
 
+    def _get_node_name_for_pod(self) -> str:
+        """Return the node name for a given pod."""
+        pod = self.client.get(Pod, name=self.unit.name.replace("/", "-"), namespace=self.namespace)
+        return pod.spec.nodeName
+
+    def get_k8s_node_ip(self) -> str:
+        """Return node IP."""
+        node = self.client.get(Node, name=self._get_node_name_for_pod(), namespace=self.namespace)
+        # [
+        #    NodeAddress(address='192.168.0.228', type='InternalIP'),
+        #    NodeAddress(address='example.com', type='Hostname')
+        # ]
+        # Remember that OpenStack, for example, will return an internal hostname, which is not
+        # accessible from the outside. Give preference to ExternalIP, then InternalIP first
+        if addr := next(
+            (a.address for a in node.status.addresses if a.type == "ExternalIP"), None
+        ):
+            return addr
+        if addr := next(
+            (a.address for a in node.status.addresses if a.type == "InternalIP"), None
+        ):
+            return addr
+        if addr := next((a.address for a in node.status.addresses if a.type == "Hostname"), None):
+            return addr
+        return None
+
+    def node_port(self, port_type="rw") -> int:
+        """Return node port."""
+        svc = self.client.get(Service, self.app.name, namespace=self.namespace)
+        if svc and svc.spec.type == "NodePort":
+            # svc.spec.ports
+            # [ServicePort(port=3306, appProtocol=None, name=None, nodePort=31438, protocol='TCP', targetPort=3306)]
+            # TODO: check if we are getting the right 3306 port
+            return svc.spec.ports[0].nodePort
+        return None
```

Comment:
In theory, it should never happen. The reason is because the `_patch_service` happens before than `reconcile_users` in the `reconcile` method. I will add a `NotImplementedError`, as this should def. be brought up to the user.

Reply:
@carlcsaposs-canonical ack, I created an exception